### ISSUE TITLE
crash: TileCokeOven cannot be cast to Reika.DragonAPI.Base.TileEntityBase

### ISSUE THREAD

**Issue by UltraVolta425** [ReikaKalseki/Reika_Mods_Issues#1416]
Hey

I recently managed to build the whole Reika's reactorcraft fusion reactor, and as soon as everything was built and the plasma was flowing nicely, I continued building on my single player map on 1.7.10. After a while, however, the game crashed and since then the map doesn't load anymore. The game starts, but as soon as I select my map in the savegame selection screen, it does start loading, but after a sec or two it crashes, with the following report:

`https://pastebin.com/j7UEWGz3`

Would you please be so kind as to look at it?

**Comment by OvermindDL1** [comment]
Hmm:
https://github.com/ReikaKalseki/DragonAPI/blob/master/Base/BlockTEBase.java#L59
```java
	public final void onNeighborChange(IBlockAccess world, int x, int y, int z, int tileX, int tileY, int tileZ)
	{
		ForgeDirection dir = ReikaDirectionHelper.getDirectionBetween(x, y, z, tileX, tileY, tileZ);
		TileEntityBase te = (TileEntityBase)world.getTileEntity(x, y, z);
		if (te != null)
			te.updateCache(dir);
	}
```
Eyup, if it receives a neighbor update from anything then it tries to just cast it straight out without confirming it is of that type first.  I'm surprised this bug has not popped up before.  Railcraft Coke ovens do 'alert' neighbor blocks as they load so the multiblock can re-form, so it makes sense as to why it is happening...  I'm just really surprised it has not been reported before...  Hmm, has it?

If you can MCEdit your world and remove your coke oven (or move it elsewhere) then the game should load fine though.  Or if it is important then I can bytecode edit a fix into the file until Reika can fix it for you (though I'd prefer not to do that, bit painful of work)?

**Comment by UltraVolta425** [comment]
Hmm, I've had sooo many crashes lately, that everytime I fear I'm about to do something close to the fusion reactor that might make the game crash, I always save the game, exit, make a backup copy of my savegame and then continue.

The backup copy of my savegame still works, so an mcedit won't be necessary since all that's missing since my backup copy and my crashed savegame is a few reinforced glass pieces :)

For safety's sake, I suppose it would be wise to remove the cake oven now?
And also, do I need to do anything to make this issue known to Reika? So that he can improve his code for another release in the future?

**Comment by UltraVolta425** [comment]
Update: so as I said in my previous post, the game loaded fine with the backup copy.
However, when I went near the place where my coke oven is, the game crashed again, with a different crash report:

`https://pastebin.com/wMSj8AUf`

Would you kindly look at this new log please? Or do I need to create a separate issue for this?

**Comment by UltraVolta425** [comment]
I'll Close this issue, seeing as this is - as far as I'm aware - a different issue.

**Comment by ReikaKalseki** [comment]
> Eyup, if it receives a neighbor update from anything then it tries to just cast it straight out without confirming it is of that type first. I'm surprised this bug has not popped up before.

That method - like most others in `Block` - assumes that it is being passed its own coordinates, i.e. some variant of `World.getBlock(x,y,z).onNeighborChange(x, y, z)`. Railcraft is failing to do this, probably passing incorrect coordinates or calling it on the wrong block.

**Comment by OvermindDL1** [comment]
> That method - like most others in Block - assumes that it is being passed its own coordinates, i.e. some variant of World.getBlock(x,y,z).onNeighborChange(x, y, z). Railcraft is failing to do this, probably passing incorrect coordinates or calling it on the wrong block.

I could have *sworn* that vanilla MC has a case or two where it calls it with other coordinates, vars were swapped or something but its been that way since Alpha days (I think it was in chunk loading code as I recall?), as a consequence I always handled it myself.  With MC programming it is better to be defensive or crashes *will* happen.  ^.^;

**Comment by UltraVolta425** [comment]
I've resolved the issue.
It's funny how often this - as a professional software tester - happens to me, that I end up resolving a bug myself. :p

The problem was an old fusion reactor on the same map I made more than a year ago. But somehow, over time, it became very buggy. In fact the whole area around it became so buggy that if you ever went near there, the game would crash. Might have been due to bad updates on my part from reika's addons, or of optifine, or maybe I MCedited a bit too much there, or maybe it was due to me terraforming the whole game from superflat to a more realistic world with mountains etc (except for the area around the old fusion reactor), I don't remember exactly.

I still had a ChickenBones Chunk Loader there.

I removed the Chunk Loader mod from the game (still kept the core and NEI though), and as of then, I haven't had any problems anymore. As a professional tester, I think it's safe to say that after two days of no errors, the bug is fixed.
